### PR TITLE
Added linspace

### PR DESCRIPTION
**Pull request by abeleinin**
## Proposed changes

Adds `mx.linspace` support based on the [numpy.linspace](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html) interface. Open to changing the interface if you'd rather imitate [torch.linspace](https://pytorch.org/docs/stable/generated/torch.linspace.html).

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/src/ops.cpp [hunk 1]
```diff
@@ -1157,6 +1157,45 @@ void init_ops(py::module_& m) {
...
+            scalar_to_double(start),
+            scalar_to_double(stop),
+            scalar_to_double(num),
+            dtype,
+            s);
+      },
+      "start"_a,
+      "stop"_a,
+      "num"_a = 50,
+      "dtype"_a = none,
```
Comment (dc-dc-dc) [inline python/src/ops.cpp:1179]: Can set dtype to float32 instead of none

File: python/src/ops.cpp [hunk 2]
```diff
@@ -1157,6 +1157,43 @@ void init_ops(py::module_& m) {
...
+            dtype_.value(),
+            s);
+      },
+      "start"_a,
+      "stop"_a,
+      "num"_a = 50,
+      "dtype"_a = float32,
+      "stream"_a = none,
+      R"pbdoc(
+      linspace(start, stop, num: Optional[float] = 50, dtype: Optional[Dtype] = float32, stream: Union[None, Stream, Device] = None) -> array
```
Comment (awni) [inline python/src/ops.cpp:1180]: Optional[int] for `num`

File: python/src/ops.cpp [hunk 3]
```diff
@@ -1157,6 +1157,43 @@ void init_ops(py::module_& m) {
...
+      "linspace",
+      [](Scalar start,
+         Scalar stop,
+         Scalar num,
+         std::optional<Dtype> dtype_,
+         StreamOrDevice s) {
+        return linspace(
+            scalar_to_double(start),
+            scalar_to_double(stop),
+            scalar_to_double(num),
```
Comment (awni) [inline python/src/ops.cpp:1170]: I think `num` here should just be an `int` as in `numpy`.

File: python/src/ops.cpp [hunk 4]
```diff
@@ -1157,6 +1157,43 @@ void init_ops(py::module_& m) {
         This can lead to unexpected results for example if `start + step`
         is a fractional value and the `dtype` is integral.
       )pbdoc");
+  m.def(
+      "linspace",
+      [](Scalar start,
+         Scalar stop,
+         Scalar num,
```
Comment (awni) [inline python/src/ops.cpp:1164]: `num` here should just be an `int`.

File: mlx/ops.h [hunk 5]
```diff
@@ -36,6 +36,17 @@ array arange(int start, int stop, int step, StreamOrDevice s = {});
...
 
+/** A 1D array of evenly spaced numbers starting at `start`
+ * stopping at `stop`, stepping by (`end` - `start`) / (`num` - 1) **/
+array linspace(
+    double start,
+    double stop,
+    int num,
+    Dtype dtype,
+    StreamOrDevice s = {});
+array linspace(double start, double stop, int num, StreamOrDevice s = {});
```
Comment (awni) [inline mlx/ops.h:47]: I think you could have `dtype = float32` and remove the overload for defaulting the dtype.

File: mlx/ops.h [hunk 6]
```diff
@@ -36,6 +36,17 @@ array arange(int start, int stop, int step, StreamOrDevice s = {});
...
+array linspace(double start, double stop, StreamOrDevice s = {});
```
Comment (awni) [inline mlx/ops.h:48]: Add a dtype with a default here as well please.

File: python/tests/test_ops.py [hunk 7]
```diff
@@ -1402,6 +1402,15 @@ def test_stack(self):
...
         self.assertTrue(np.array_equal(c, np_c))
 
+    def test_linspace(self):
+        a = mx.linspace(0, 10, 5)
+        np_a = np.linspace(0, 10, 5)
+        self.assertTrue(np.array_equal(a, np_a))
+
+        b = mx.linspace(0, 10, 5, dtype=mx.int32)
+        np_b = np.linspace(0, 10, 5, dtype=int)
+        self.assertTrue(np.array_equal(b, np_b))
```
Comment (awni) [inline python/tests/test_ops.py:1412]: Could you add a few more tests here, maybe some where the `start` and `stop` have unusual types, or edge cases like `num=0`?

**Review by awni — COMMENTED** [review]
Really nice, thanks for adding this it will be super useful! I left a few comments, please address and then we can merge.

Aslo could you add `linspace` to the [docs](https://github.com/ml-explore/mlx/blob/main/docs/src/python/ops.rst)?

**Review by awni — CHANGES_REQUESTED** [review]
Sorry meant to request changes.

File: python/tests/test_ops.py (hunk 7 above)
Comment (abeleinin) [reply python/tests/test_ops.py:1412]: Sorry for the delay getting this merged. I ran into an issue testing irrational number step sizes. For example:

Screenshot 2023-12-18 at 2 37 00 AM

After some digging, I believe this is an issue with the float precision of `mx.arange` (limitation of numbers being stored as `float32`).

Here's an example to show the floating point issue with `mx.arange`:

Screenshot 2023-12-18 at 2 38 30 AM

I also tested against `torch.linspace` and it has similar weird floating point behavior, but the outputs are not consistent.

Screenshot 2023-12-18 at 3 02 22 AM

Maybe we need to add `float64` support to `mx.arange`? Let me know your thoughts.

Comment (awni) [reply python/tests/test_ops.py]: @abeleinin that's a good point but at the moment we don't support `mx.float64` and we probably won't have it for a little while. So for testing you can just use `mx.allclose` instead of `array_equal` to make sure values are sufficiently similar.

**Comment by awni** [comment]
@dc-dc-dc Let me know when you are ready for re-review. I think you need to rebase on main as well to resolve conflicts.

File: python/tests/test_ops.py [hunk 8]
```diff
@@ -1402,6 +1402,27 @@ def test_stack(self):
         np_c = np.stack([np_a, np_b], axis=1)
         self.assertTrue(np.array_equal(c, np_c))
 
+    def test_linspace(self):
+        # Test default num = 50
+        a = mx.linspace(0, 1)
```
Comment (dc-dc-dc) [inline python/tests/test_ops.py:1407]: This test case is failing, getting a shape of size 51 and there is an extra step past one at the end 
`
array([0, 0.0204082, 0.0408163, ..., 0.979592, 1, 1.02041], dtype=float32)
`

File: python/tests/test_ops.py [hunk 9]
```diff
@@ -1402,6 +1402,27 @@ def test_stack(self):
...
+        np_a = np.linspace(0, 1)
+        self.assertTrue(np.array_equal(a, np_a))
+
+        # Test int32 dtype
+        b = mx.linspace(0, 10, 5, dtype=mx.int32)
+        np_b = np.linspace(0, 10, 5, dtype=int)
+        self.assertTrue(np.array_equal(b, np_b))
```
Comment (dc-dc-dc) [inline python/tests/test_ops.py:1414]: can you utilize, `self.assertEqualArray(b, mx.array(np_b))`
It will check
- shape
- dtype
- correctness of result with options for atol/rtol

File: python/tests/test_ops.py [hunk 10]
```diff
@@ -1402,6 +1402,27 @@ def test_stack(self):
...
+
+        # Test negative sequence with float start and stop
+        c = mx.linspace(-2.7, -0.7, 7)
+        np_c = np.linspace(-2.7, -0.7, 7)
+        self.assertTrue(np.array_equal(c, np_c))
```
Comment (dc-dc-dc) [inline python/tests/test_ops.py:1419]: negative test cases for int32 don't match numpy

```python
>>> np.linspace(-10, -1, dtype=np.int32)
array([-10, -10, -10, -10, -10, -10,  -9,  -9,  -9,  -9,  -9,  -8,  -8,
        -8,  -8,  -8,  -8,  -7,  -7,  -7,  -7,  -7,  -6,  -6,  -6,  -6,
        -6,  -6,  -5,  -5,  -5,  -5,  -5,  -4,  -4,  -4,  -4,  -4,  -4,
        -3,  -3,  -3,  -3,  -3,  -2,  -2,  -2,  -2,  -2,  -1], dtype=int32)
>>> mx.linspace(-10, -1, dtype=mx.int32)
array([-10, -9, -9, ..., -1, -1, 0], dtype=int32)
```

**Comment by abeleinin** [comment]
@awni Getting this finished up now!

**Review by awni — APPROVED** [review]
This is awesome!! Thanks for adding it. I added some minor stylistic changes and I'm merging it now.

**Comment by awni** [comment]
PS @dc-dc-dc thanks for the reviews :)
